packages/xrpc-server/src/stream/websocket-keepalive.ts

```diff
@@ -0,0 +1,146 @@
+  public isAlive = false
+  public heartbeatInterval: NodeJS.Timer | null
+
+  constructor(
+    public opts: ClientOptions & {
+      getUrl: () => Promise<string>
+      maxReconnectSeconds?: number
+      signal?: AbortSignal
+      heartbeatInterval?: number
+      onReconnectError?: (
+        error: unknown,
+        n: number,
+        initialSetup: boolean,
+      ) => void
+    },
+  ) {}
+
+  async *[Symbol.asyncIterator](): AsyncGenerator<Uint8Array> {
+    const maxReconnectMs = 1000 * (this.opts.maxReconnectSeconds ?? 64)
+    while (true) {
+      if (this.reconnects !== null) {
+        const duration = this.initialSetup
+          ? Math.min(1000, maxReconnectMs)
+          : backoffMs(this.reconnects++, maxReconnectMs)
+        await wait(duration)
+      }
+      const url = await this.opts.getUrl()
+      this.ws = new WebSocket(url, this.opts)
+      const ac = new AbortController()
+      if (this.opts.signal) {
+        forwardSignal(this.opts.signal, ac)
+      }
+      this.ws.once('open', () => {
+        this.initialSetup = false
+        this.reconnects = 0
+      })
+      this.ws.once('close', (code, reason) => {
+        if (code === CloseCode.Abnormal) {
+          // Forward into an error to distinguish from a clean close
+          ac.abort(
+            new AbnormalCloseError(`Abnormal ws close: ${reason.toString()}`),
+          )
+        }
+        this.clearHeartbeatInterval()
+      })
+
+      // send a ping on an interval and check to ensure that a pong has come back in that time
+      this.clearHeartbeatInterval() // clean up any previous interval that's stuck around
+      this.isAlive = true
+      // send a ping every x seconds
+      this.heartbeatInterval = setInterval(() => {
+        if (!this.isAlive) {
+          this.ws?.terminate()
+        }
+        this.isAlive = false
+        this.ws?.ping()
+      }, this.opts.heartbeatInterval ?? 10 * SECOND)
+
+      // add a heartbeat
+      this.ws.on('pong', () => {
+        this.isAlive = true
+      })
+
+      try {
+        const wsStream = streamByteChunks(this.ws, { signal: ac.signal })
+        for await (const chunk of wsStream) {
+          // @TODO asdf
+          yield chunk
+        }
+      } catch (_err) {
+        const err = _err?.['code'] === 'ABORT_ERR' ? _err['cause'] : _err
+        if (err instanceof DisconnectError) {
+          // We cleanly end the connection
+          this.ws?.close(err.wsCode)
+          break
+        }
+        this.ws?.close() // No-ops if already closed or closing
+        if (isReconnectable(err)) {
+          this.reconnects ??= 0 // Never reconnect with a null
+          this.opts.onReconnectError?.(err, this.reconnects, this.initialSetup)
+          continue
+        } else {
+          throw err
+        }
+      }
+      break // Other side cleanly ended stream and disconnected
+    }
+  }
+
+  clearHeartbeatInterval() {
+    if (this.heartbeatInterval === null) return
```

Comment:
`heartbeatInterval` isn't initialized to `null`, so this condition might not get triggered when it's intended to.  We could initialize `heartbeatInterval` or make this check looser, but I also don't really think the check is needed if we want to just nix it.

Reply:
ah good catch 🙏